oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -0,0 +1,122 @@
+ * token for STS exchange.
+ */
+public class CertificateIdentityPoolSubjectTokenSupplier
+    implements IdentityPoolSubjectTokenSupplier {
+
+  private static final Gson GSON = new Gson();
+  private final IdentityPoolCredentialSource credentialSource;
+
+  CertificateIdentityPoolSubjectTokenSupplier(IdentityPoolCredentialSource credentialSource) {
+    this.credentialSource = checkNotNull(credentialSource, "credentialSource cannot be null");
+    // This check ensures that the credential source was intended for certificate usage.
+    // IdentityPoolCredentials logic should guarantee credentialLocation is set in this case.
+    checkNotNull(
+        credentialSource.certificateConfig,
+        "credentialSource.certificateConfig cannot be null when creating"
+            + " CertificateIdentityPoolSubjectTokenSupplier");
+  }
+
+  private static X509Certificate loadLeafCertificate(String path)
+      throws IOException, CertificateException {
+    byte[] leafCertBytes;
+    try {
+      // IdentityPoolCredentials should have already validated the path exists via X509Provider.
+      leafCertBytes = Files.readAllBytes(Paths.get(path));
+    } catch (InvalidPathException e) {
+      throw new IOException("Invalid certificate file path provided: " + path, e);
+    }
+    // Files.readAllBytes throws IOException for other read errors.
+    return parseCertificate(leafCertBytes);
+  }
+
+  private static X509Certificate parseCertificate(byte[] certData) throws CertificateException {
+    if (certData == null || certData.length == 0) {
+      throw new IllegalArgumentException("Invalid certificate data: empty or null input");
+    }
+
+    try {
+      CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+      InputStream certificateStream = new ByteArrayInputStream(certData);
+      return (X509Certificate) certificateFactory.generateCertificate(certificateStream);
+    } catch (CertificateException e) {
+      throw new CertificateException("Failed to parse X.509 certificate data.", e);
+    }
+  }
+
+  private static String encodeCert(X509Certificate certificate)
+      throws CertificateEncodingException {
+    return Base64.getEncoder().encodeToString(certificate.getEncoded());
+  }
+
+  @Override
+  public String getSubjectToken(ExternalAccountSupplierContext context) throws IOException {
+    try {
+      // credentialSource.credentialLocation is expected to be non-null here,
+      // set during IdentityPoolCredentials construction for certificate type.
+      X509Certificate leafCert = loadLeafCertificate(credentialSource.credentialLocation);
+      String encodedCert = encodeCert(leafCert);
+
+      JsonArray certChain = new JsonArray();
+      certChain.add(new JsonPrimitive(encodedCert));
+
+      return GSON.toJson(certChain);
+    } catch (CertificateException e) {
+      throw new IOException(
+          "Failed to parse certificate from: " + credentialSource.credentialLocation, e);
+    }
```

Comment:
same here, is there any specific reasons some exceptions are re-thrown as IOExceptions?

Reply:
The main reason we catch CertificateException and re-throw it as an IOException here is to comply with the `IdentityPoolSubjectTokenSupplier` interface, which declares that `getSubjectToken()` throws IOException.

Reply:
Ah understood, can you add a small comment with the above info?